fix(ngcc): capture UMD/CommonJS inner class implementation nodes correctly

petebacondarwin opened the pull request:
Previously, UMD/CommonJS class inline declarations of the form:

```ts
exports.Foo = (function() { function Foo(); return Foo; })();
```

were capturing the whole IIFE as the implementation, rather than
the inner class (i.e. `function Foo() {}` in this case). This caused
the interpreter to break when it was trying to access such an export,
since it would try to evaluate the IIFE rather than treating it as a class
declaration.

petebacondarwin committed:
fix(ngcc): capture UMD/CommonJS inner class implementation node correctly

Previously, UMD/CommonJS class inline declarations of the form:

```ts
exports.Foo = (function() { function Foo(); return Foo; })();
```

were capturing the whole IIFE as the implementation, rather than
the inner class (i.e. `function Foo() {}` in this case). This caused
the interpreter to break when it was trying to access such an export,
since it would try to evaluate the IIFE rather than treating it as a class
declaration.

gkalpak commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -342,7 +342,8 @@ export class StaticInterpreter {
-    return decl.kind === DeclarationKind.Inline && decl.implementation !== undefined ?
+    return decl.kind === DeclarationKind.Inline && decl.implementation !== undefined &&
+            !isDeclaration(decl.implementation) ?
         // Inline declarations with an `implementation` should be visited as expressions
```
This comment could be updated to mention the new requirement or `implementation` being a `ts.Expression`.

gkalpak commented on packages/compiler-cli/ngcc/test/host/umd_host_spec.ts:
```diff
@@ -2732,10 +2733,8 @@ runInEachFileSystem(() => {
-    return SomeClass;
-  }())`,
+                  'SomeClass',
+                  `SomeClass = (function() {\n    function SomeClass() {}\n    return SomeClass;\n  }())`,
```
```suggestion
                  'SomeClass = (function() {\n    function SomeClass() {}\n    return SomeClass;\n  }())',
```

gkalpak reviewed (approved):
LGTM with a couple of comments:
- It would be cool if we could have an integration test showing all the bits working together (unless it would be too messy).
- Commit message typo: `Fool` --> `Foo` :grin:

petebacondarwin committed:
fixup! fix(ngcc): capture UMD/CommonJS inner class implementation node correctly

petebacondarwin committed:
fixup! fix(ngcc): capture UMD/CommonJS inner class implementation node correctly

petebacondarwin committed:
fixup! fix(ngcc): capture UMD/CommonJS inner class implementation node correctly

petebacondarwin committed:
fixup! fix(ngcc): capture UMD/CommonJS inner class implementation node correctly

petebacondarwin commented:
Hmm, an integration test... Let me look into that. Trouble is that we don't have any UMD packages to test against.

alxhub commented:
[Presubmit](http://test/OCL:338278630:BASE:338278645:1603297845685:127d159e)

petebacondarwin commented:
@gkalpak - I managed to create an integration test that actually fails without this fix!!

JoostK commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -467,6 +467,58 @@ runInEachFileSystem(() => {
        });
 
+    it('should support inline UMD/CommonJS exports declarations', () => {
+      // Setup an Angular entry-point in UMD module format has an inline exports declaration
```
```suggestion
      // Setup an Angular entry-point in UMD module format which has an inline exports declaration
```

petebacondarwin replied:
I'm going with `that` instead but 🤷

JoostK commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -467,6 +467,58 @@ runInEachFileSystem(() => {
+            exports.FooModule = /** @class */ (function () {
+              function FooModule() {}
+              FooModule = __decorate([
+                  core.NgModule({declarations: [exports.FooDirective]})
```
You can make this more challenging when this is a reference to an inline declaration of an expression.
```suggestion
                  core.NgModule({declarations: exports.declarations})
```

and then have

```ts
exports.declarations = [exports.FooDirective];
```

this was the pattern that broke after the initial fix.

petebacondarwin replied:
Great! Will do.

JoostK commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -467,6 +467,58 @@ runInEachFileSystem(() => {
+               targetEntryPointPath: 'test-package',
+               propertiesToConsider: ['main'],
+             }))
+          .not.toThrow();
```
I'd extend this to verify that the output files contain AOT output, as this wouldn't currently fail when none of the declarations are found at all.

petebacondarwin replied:
Yep - good idea

JoostK commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -517,6 +519,13 @@ runInEachFileSystem(() => {
+          .toContain('FooModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FooModule });');
+      expect(processedFile)
+          .toContain(
+              'ɵngcc0.ɵɵsetNgModuleScope(FooModule, { declarations: function () { return [FooDirective]; } });');
```
Uhm, this looks like a bug w.r.t `FooDirective`, as that won't be declared :( We should refer to it as `exports.FooDirective`

petebacondarwin replied:
Ahah! OMG. I didn't notice....
Hmm, so we appear to be referring to inner class declaration here...
😭

petebacondarwin replied:
Right, so the problem is that in the Angular compiler we refer to classes such as `exports.FooDirective` via a `Reference` object, which gives special meaning to identifiers that "reference" the class.

Then when we are converting the `Reference` to an `R3Reference` the `LocalIdentifierStrategy` makes the assumption that any identifier of a reference is free standing (i.e. it is a single expression in its own right and not the RHS of a property access).  This is why we end up emitting `FooDirective` rather than `exports.FooDirective`.

I think the solution is to get the `LocalIdentifierStrategy` to check whether the reference identifier is actually the child of a property expression, and if so then return that parent expression rather than the identifier itself.

Anyone have any better ideas?

petebacondarwin committed:
refactor(compiler-cli): support namespaced references

The compiler uses a `Reference` abstraction to refer to TS nodes
that it needs to refer to from other parts of the source. Such
references keep track of any identifiers that represent the referenced
node.

Prior to this commit, the compiler (and specifically `ReferenceEmitter`
classes) assumed that the reference identifiers are always free standing.
In other words a reference identifier would be an expression like
`FooDirective` in the expression `class FooDirective {}`.

But in UMD/CommonJS source, a reference can actually refer to an "exports"
declaration of the form `exports.FooDirective = ...`.
In such cases the `FooDirective` identifier is not free-standing
since it is part of a property access, so the `ReferenceEmitter`
should take this into account when emitting an expression that
refers to such a `Reference`.

This commit changes the `LocalIdentifierStrategy` reference emitter
so that if the `node` being referenced is not a declaration itself and
is in the current file, then it should be used directly, rather than
trying to use one of its identifiers.

AndrewKushnir commented:
[Presubmit](http://test/OCL:338689278:BASE:338689303:1603470826992:87bc035f).

AndrewKushnir commented:
FYI, presubmit is successful (I've updated g3 status for this PR).

petebacondarwin commented:
Thanks @AndrewKushnir
